pkg/workflow/workflow.go

```diff
@@ -270,79 +290,88 @@ func (w *workflow) Trace() error {
 	return recorder.With(w.cli, w.app).Save("", data).Limit(10).Error()
 }
 
-func (w *workflow) GetBackoffWaitTime() time.Duration {
-	nextTime, ok := w.wfCtx.GetValueInMemory(wfTypes.ContextKeyNextExecuteTime)
-	if !ok {
-		return time.Second
+func (w *workflow) GetSuspendBackoffWaitTime() time.Duration {
+	if w.app.Status.Workflow == nil {
+		return 0
 	}
-	unix, ok := nextTime.(int64)
-	if !ok {
-		return time.Second
+	stepStatus := make(map[string]common.StepStatus)
+	for _, ss := range w.app.Status.Workflow.Steps {
+		setStepStatus(stepStatus, ss.StepStatus)
+		for _, sss := range ss.SubStepsStatus {
+			setStepStatus(stepStatus, sss.StepStatus)
+		}
 	}
-	next := time.Unix(unix, 0)
-	if next.After(time.Now()) {
-		return time.Until(next)
+	max := time.Duration(1<<63 - 1)
+	min := max
+	if w.app.Spec.Workflow != nil {
+		for _, step := range w.app.Spec.Workflow.Steps {
+			if step.Type == wfTypes.WorkflowStepTypeSuspend || step.Type == wfTypes.WorkflowStepTypeStepGroup {
+				min = handleSuspendBackoffTime(step, stepStatus[step.Name], min)
+			}
+			for _, sub := range step.SubSteps {
+				if sub.Type == wfTypes.WorkflowStepTypeSuspend {
+					min = handleSuspendBackoffTime(oamcore.WorkflowStep{
+						Name:       sub.Name,
+						Type:       sub.Type,
+						Timeout:    sub.Timeout,
+						Properties: sub.Properties,
+					}, stepStatus[sub.Name], min)
+				}
+			}
+		}
 	}
-
-	return time.Second
+	if min == max {
+		return 0
+	}
+	return min
 }
 
-func (w *workflow) HandleSuspendWait(ctx monitorContext.Context) (doWaiting bool, durationWaiting time.Duration, errRet error) {
-	ctx.Info("handle suspend wait")
-	for i, stepStatus := range w.app.Status.Workflow.Steps {
-		if !w.isWaitSuspendStep(stepStatus) {
-			continue
-		}
-
-		step := w.getWorkflowStepByName(stepStatus.Name)
-		if step.Name == "" {
-			errRet = fmt.Errorf("failed to get workflow step by name: %s", stepStatus.Name)
-			return
-		}
-
-		d, wd, err := wfTasks.GetSuspendStepDurationWaiting(step)
-		if err != nil {
-			ctx.Error(err, "failed to get suspend step wait duration")
-			errRet = err
-			return
-		}
-
-		if d {
-			doWaiting = d
-			remainingDuration := wd - time.Since(stepStatus.FirstExecuteTime.Time)
-			if remainingDuration <= 0 {
-				w.app.Status.Workflow.Steps[i].Phase = common.WorkflowStepPhaseSucceeded
+func handleSuspendBackoffTime(step oamcore.WorkflowStep, status common.StepStatus, min time.Duration) time.Duration {
+	if status.Phase == common.WorkflowStepPhaseRunning {
```

Comment:
please return the result first to avoid code block complex:

```
if status.Phase != common.WorkflowStepPhaseRunning {
   return min
}
...
// the rest code can avoid indentation
```

Reply:
In this way, the rest code can avoid indentation indeed, but we have to write `return min` twice, so I think there's no necessity to avoid indentation.

Reply:
if so, we should make every function which has an error return like this:

```
if err==nil{
   ...
   ...
}

return err
```

This can avoid multiple `return err` block, but nobody actually did like this.

Reply:
refer to the go code convention which is also kubevela's code convention https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow

Reply:
Just in this case, there are not many differences between:
```go
if status.Phase != common.WorkflowStepPhaseRunning {
   return min
}
.... // some codes
return min
```
and 
```go
if status.Phase == common.WorkflowStepPhaseRunning {
.... // some codes
}
return min
```

And I think the last one is better. But in some cases, there's better to return the result first.